x/ibc-rate-limit/ibc_middleware.go

```diff
@@ -0,0 +1,320 @@
+)
+
+var (
+	_ porttypes.Middleware  = &IBCModule{}
+	_ porttypes.ICS4Wrapper = &ICS4Middleware{}
+)
+
+type ICS4Middleware struct {
+	channel        porttypes.ICS4Wrapper
+	accountKeeper  *authkeeper.AccountKeeper
+	BankKeeper     *bankkeeper.BaseKeeper
+	ContractKeeper *wasmkeeper.PermissionedKeeper
+	LockupKeeper   *lockupkeeper.Keeper
+	ParamSpace     paramtypes.Subspace
+}
+
+func NewICS4Middleware(
+	channel porttypes.ICS4Wrapper,
+	accountKeeper *authkeeper.AccountKeeper, contractKeeper *wasmkeeper.PermissionedKeeper,
+	bankKeeper *bankkeeper.BaseKeeper, lockupKeeper *lockupkeeper.Keeper,
+	paramSpace paramtypes.Subspace,
+) ICS4Middleware {
+	return ICS4Middleware{
+		channel:        channel,
+		accountKeeper:  accountKeeper,
+		ContractKeeper: contractKeeper,
+		BankKeeper:     bankKeeper,
+		LockupKeeper:   lockupKeeper,
+		ParamSpace:     paramSpace,
+	}
+}
+
+// SendPacket implements the ICS4 interface and is called when sending packets.
+// This method retrieves the contract from the middleware's parameters and checks if the limits have been exceeded for
+// the current transfer, in which case it returns an error preventing the IBC send from taking place.
+// If the contract param is not configured, or the contract doesn't have a configuration for the (channel+denom) being
+// used, transfers are not prevented and handled by the wrapped IBC app
+func (i *ICS4Middleware) SendPacket(ctx sdk.Context, chanCap *capabilitytypes.Capability, packet exported.PacketI) error {
+	contract := i.GetParams(ctx)
+	if contract == "" {
+		// The contract has not been configured. Continue as usual
+		return i.channel.SendPacket(ctx, chanCap, packet)
+	}
+
+	amount, denom, err := GetFundsFromPacket(packet)
+	if err != nil {
+		return sdkerrors.Wrap(err, "Rate limited SendPacket")
+	}
+	channelValue := i.CalculateChannelValue(ctx, denom)
+	err = CheckAndUpdateRateLimits(
+		ctx,
+		i.ContractKeeper,
+		"send_packet",
+		contract,
+		channelValue,
+		packet.GetSourceChannel(),
+		denom,
+		amount,
+	)
+	if err != nil {
+		return sdkerrors.Wrap(err, "Rate limited SendPacket")
+	}
+
+	return i.channel.SendPacket(ctx, chanCap, packet)
+}
+
+func (i *ICS4Middleware) WriteAcknowledgement(ctx sdk.Context, chanCap *capabilitytypes.Capability, packet exported.PacketI, ack exported.Acknowledgement) error {
+	return i.channel.WriteAcknowledgement(ctx, chanCap, packet, ack)
+}
+
+func (i *ICS4Middleware) GetParams(ctx sdk.Context) (contract string) {
+	i.ParamSpace.GetIfExists(ctx, []byte("contract"), &contract)
+	return contract
+}
+
+// CalculateChannelValue The value of an IBC channel. This is calculated using the denom supplied by the sender.
+// if the denom is not correct, the transfer should fail somewhere else on the call chain
+func (i *ICS4Middleware) CalculateChannelValue(ctx sdk.Context, denom string) sdk.Int {
+	return i.BankKeeper.GetSupplyWithOffset(ctx, denom).Amount
+}
+
+type IBCModule struct {
```

Comment:
Wdyt about defining this in a separate file `ibc_module.go`?

Noticed the suggestion here:
https://github.com/cosmos/ibc-go/blob/f57170b1d4dd202a3c6c1c61dcf302b6a9546405/docs/ibc/middleware/develop.md?plain=1#L58

which seems to help split the abstractions per file

Reply:
sure! I find it more clear to have the whole middleware in the same file, but no strong preference, so I'll separate them